File: src/test/java/com/automation/framework/utils/ExcelUtils.java
package com.automation.framework.utils;

import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;

public final class ExcelUtils {

  private ExcelUtils() {}

  public static Object[][] readSheet(Path file, String sheetName) {
    try (InputStream is = Files.newInputStream(file);
         Workbook wb = WorkbookFactory.create(is)) {

      Sheet sheet = wb.getSheet(sheetName);
      if (sheet == null) {
        throw new IllegalArgumentException("Sheet not found: " + sheetName);
      }

      int lastRow = sheet.getLastRowNum();
      if (lastRow < 1) {
        return new Object[0][0];
      }

      Row header = sheet.getRow(0);
      int cols = header.getLastCellNum();

      // Data rows start at row 1
      Object[][] data = new Object[lastRow][cols];
      DataFormatter fmt = new DataFormatter();

      for (int r = 1; r <= lastRow; r++) {
        Row row = sheet.getRow(r);
        for (int c = 0; c < cols; c++) {
          Cell cell = row == null ? null : row.getCell(c);
          data[r - 1][c] = cell == null ? "" : fmt.formatCellValue(cell);
        }
      }
      return data;

    } catch (Exception e) {
      throw new RuntimeException("Failed reading excel: " + file, e);
    }
  }

  public static void ensureSampleLoginWorkbook(Path file) {
    if (Files.exists(file)) {
      return;
    }

    try {
      Files.createDirectories(file.getParent());

      try (Workbook wb = new XSSFWorkbook()) {
        Sheet sheet = wb.createSheet("Login");

        Row header = sheet.createRow(0);
        header.createCell(0).setCellValue("testName");
        header.createCell(1).setCellValue("username");
        header.createCell(2).setCellValue("password");
        header.createCell(3).setCellValue("expectedContains");
        header.createCell(4).setCellValue("run");

        Row r1 = sheet.createRow(1);
        r1.createCell(0).setCellValue("valid_login");
        r1.createCell(1).setCellValue("tomsmith");
        r1.createCell(2).setCellValue("SuperSecretPassword!");
        r1.createCell(3).setCellValue("You logged into a secure area!");
        r1.createCell(4).setCellValue("Y");

        Row r2 = sheet.createRow(2);
        r2.createCell(0).setCellValue("invalid_login");
        r2.createCell(1).setCellValue("wrong");
        r2.createCell(2).setCellValue("wrong");
        r2.createCell(3).setCellValue("Your username is invalid!");
        r2.createCell(4).setCellValue("Y");

        // Apache POI auto-sizing can require a graphical environment (fonts/X11),
        // which is not available on many CI/headless Linux environments.
        // Use fixed column widths instead to keep sample generation robust.
        int[] widths = {
            18 * 256, // testName
            16 * 256, // username
            22 * 256, // password
            40 * 256, // expectedContains
            6 * 256   // run
        };
        for (int c = 0; c < widths.length; c++) {
          sheet.setColumnWidth(c, widths[c]);
        }

        try (var os = Files.newOutputStream(file)) {
          wb.write(os);
        }
      }
    } catch (Exception e) {
      throw new RuntimeException("Failed creating sample workbook: " + file, e);
    }
  }
}

File: python/framework/utils/excel_utils.py
from pathlib import Path

from openpyxl import Workbook, load_workbook


def read_sheet(file_path, sheet_name):
    path = Path(file_path)
    if not path.exists():
        raise RuntimeError(f"Excel file not found: {path}")

    wb = load_workbook(path, data_only=True)
    try:
        if sheet_name not in wb.sheetnames:
            raise RuntimeError(f"Sheet not found: {sheet_name}")

        sheet = wb[sheet_name]
        rows = list(sheet.iter_rows(values_only=True))
        if len(rows) < 2:
            return []

        header = rows[0]
        cols = len(header)
        data = []

        for row in rows[1:]:
            row_values = []
            for idx in range(cols):
                cell = row[idx] if idx < len(row) else None
                row_values.append("" if cell is None else str(cell))
            data.append(row_values)

        return data
    finally:
        wb.close()


def ensure_sample_login_workbook(file_path):
    path = Path(file_path)
    if path.exists():
        return

    path.parent.mkdir(parents=True, exist_ok=True)

    wb = Workbook()
    sheet = wb.active
    sheet.title = "Login"

    header = ["testName", "username", "password", "expectedContains", "run"]
    sheet.append(header)

    sheet.append(
        ["valid_login", "tomsmith", "SuperSecretPassword!", "You logged into a secure area!", "Y"]
    )
    sheet.append(["invalid_login", "wrong", "wrong", "Your username is invalid!", "Y"])

    wb.save(path)
